File: src/scrape.ts
import cheerio from 'cheerio';
import axios, { AxiosResponse } from 'axios';

interface Server {
  status: string;
  category: string;
  characterCreationStatus: string;
}

interface DataCentre {
  [name: string]: Server;
}

/**
 * Object that maps names of data centres to their servers
 */
interface Servers {
  [name: string]: DataCentre;
}

/**
 * Fetches the server status page.
 * @returns The fetch response if it was successful.
 */
async function fetchStatus(): Promise<AxiosResponse<any, any>> {
  const resp = await axios.get(
    'https://na.finalfantasyxiv.com/lodestone/worldstatus/',
    { responseType: 'text' }
  );

  if (resp.status === 200) {
    return resp;
  }

  throw new Error(`Could not fetch world status page. Reason: ${resp.status}`);
}

/**
 * Parses a server <li> element and retrieves relevant attributes.
 * @param {cheerio.Element} server
 * @param {cheerio.Root} $
 * @returns Object containing a particular server's status information
 */
function getServerInfo($: cheerio.Root, server: cheerio.Element): Server {
  const selector: cheerio.Cheerio = $(server).parent();

  const category: string = selector
    .find('div .world-list__world_category')
    .find('p')
    .text();

  const status: string = selector
    .find('div .world-list__status_icon')
    .find('i')
    .attr('data-tooltip')!
    .trim();

  const characterCreationStatus: string = selector
    .find('div .world-list__create_character')
    .find('i')
    .attr('data-tooltip')!;

  return {
    status: status,
    category: category,
    characterCreationStatus: characterCreationStatus,
  };
}

/**
 * Scrapes the FFXIV Server Status page using Cheerio.
 * @returns Object containing status information for each server, organized by which data centre they belong to
 */
export async function scrapeServerStatus(): Promise<Servers> {
  const servers: Servers = Object();

  const resp = await fetchStatus();
  const $ = cheerio.load(resp.data);

  // Information for each server is nested in an
  // <li> element named `item-list`
  $('.world-dcgroup__item').each((_i: number, e: cheerio.Element) => {
    const dataCentre: string = $(e).find('h2').text();

    servers[dataCentre] = Object();

    $(e)
      .find('.world-list__world_name')
      .each((_i: number, e: cheerio.Element) => {
        servers[dataCentre][$(e).find('p').text()] = getServerInfo($, e);
      });
  });

  return servers;
}

File: internal/lodestone/scrape.go
// Package lodestone scrapes FFXIV world-status information from the Lodestone website.
package lodestone

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
)

// ServerStatus describes a single world's current state.
type ServerStatus struct {
	Status                  string // "Online" | "Offline"
	Category                string // "Standard" | "Preferred" | "Congested" | "New"
	CharacterCreationStatus string // "Creation of New Characters Available" | "...Unavailable"
}

// Servers maps a world name (e.g. "Adamantoise") to its current status.
type Servers map[string]ServerStatus

// Scraper fetches and parses the Lodestone world-status page.
type Scraper struct {
	httpClient *http.Client
	url        string
}

// NewScraper builds a Scraper targeting the given world-status URL. If httpClient is nil,
// a client with a sane default timeout is used.
func NewScraper(httpClient *http.Client, url string) *Scraper {
	if httpClient == nil {
		httpClient = &http.Client{Timeout: 15 * time.Second}
	}

	return &Scraper{httpClient: httpClient, url: url}
}

// Scrape fetches the world-status page and parses it into a Servers map.
func (scraper *Scraper) Scrape(ctx context.Context) (Servers, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, scraper.url, nil)
	if err != nil {
		return nil, fmt.Errorf("build world status request: %w", err)
	}

	resp, err := scraper.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("fetch world status page: %w", err)
	}
	// A close error on an already-fully-read response body carries no useful signal here.
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("could not fetch world status page: status %d", resp.StatusCode)
	}

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("parse world status page: %w", err)
	}

	servers := Servers{}

	doc.Find("li .item-list").Each(func(_ int, selection *goquery.Selection) {
		name := strings.TrimSpace(selection.Find("div .world-list__world_name p").Text())
		if name == "" {
			return
		}

		servers[name] = parseServerStatus(selection)
	})

	// A 200 response that parses to zero worlds almost always means Lodestone changed
	// its markup and our selectors no longer match, rather than every world vanishing.
	// Returning an error here routes the failure through the same path as a network
	// error, so the monitor preserves its last-known-good snapshot instead of wiping it.
	if len(servers) == 0 {
		return nil, fmt.Errorf("parsed zero worlds from world status page (layout may have changed)")
	}

	return servers, nil
}

// parseServerStatus extracts a single world's status fields from its <li> element.
func parseServerStatus(selection *goquery.Selection) ServerStatus {
	category := strings.TrimSpace(selection.Find("div .world-list__world_category p").Text())

	status := strings.TrimSpace(
		selection.Find("div .world-list__status_icon i").AttrOr("data-tooltip", ""),
	)

	characterCreationStatus := strings.TrimSpace(
		selection.Find("div .world-list__create_character i").AttrOr("data-tooltip", ""),
	)

	return ServerStatus{
		Status:                  status,
		Category:                category,
		CharacterCreationStatus: characterCreationStatus,
	}
}
